dhall/src/Dhall/Pretty/Internal.hs

```diff
@@ -1072,14 +1073,30 @@ prettyCharacterSet characterSet expression =
         angles . map prettyAlternative . Dhall.Map.toList
 
     prettyChunks :: Pretty a => Chunks Src a -> Doc Ann
-    prettyChunks (Chunks a b) =
-        if anyText (== '\n')
-        then
+    prettyChunks (Chunks a b)
+        | anyText (\c -> Data.Char.isControl c && not (Data.Char.isSpace c)) =
+            short
+        | leadingWhitespace =
+            short
+        | anyText (== '\n') =
             if anyText (/= '\n')
             then long
             else Pretty.flatAlt long short
-        else short
+        | otherwise =
+            short
       where
+        anyText predicate = any (\(text, _) -> Text.any predicate text) a || Text.any predicate b
+
+        -- If we format the text literal as a multi-line string, would any line
```

Comment:
Multi-line `Text` literals do not preclude leading semantically-significant whitespace.  For example:

```dhall
''
 a
b
''
```

Reply:
Oh, I think I was confused by the [`leadingTabs` test case](https://github.com/dhall-lang/dhall-lang/blob/master/tests/parser/success/leadingTabsA.dhall) then.

Does the explanation on that test case simply mean that tabs are treated just like spaces for dedenting purposes?

Reply:
And why does this typecheck?

```dhall
assert :
''

 ''
===
"\n"
```

```
$ dhall --file spaces.dhall 
assert : "\n" ≡ "\n"
```

Reply:
@sjakobi: Yes, dedenting will strip the longest shared prefix of tabs and/or spaces for each line

The reason that example type-checks is because leading tabs/spaces before the opening `''` do not count for dedenting purposes, so the dedenter only sees two line prefixes, both of which begin with two spaces, so those two-space prefixes are stripped, leaving behind only a newline.

Reply:
Hmm, I don't get it. Maybe this clearer:

```
$ dhall format
''
 ''
<Ctrl-D>
""
```

There's no whitespace before the opening `''`, so the leading space before the closing `''` shouldn't actually be subtracted, right?

Reply:
@sjakobi: The whitespace before the opening `''` is ignored for the purposes of computing the longest shared prefix.  So the only whitespace prefix is the one right before the closing `''` and that prefix is stripped because there are no other whitespaces prefixes to compare to, therefore it must be the longest whitespace prefix.

Reply:
@sjakobi: I highly recommend reading this section of the standard: https://github.com/dhall-lang/dhall-lang/blob/master/standard/multiline.md#indentation

Reply:
Thank you @Gabriel439. I had never read that document and somehow forgot that it existed at all. Maybe we should add a link to it from the grammar?

So if I understood correctly we must only avoid multi-line formatting if the _last line_ has leading whitespace. Then this would indeed be a bug:

```
$ echo '"\n "' | dhall format
''

 ''
```